R/assert.R

```diff
@@ -2,13 +2,41 @@
 #'
 #' @description
 #'
-#' * `vec_is()` is a predicate that checks if its input conforms to a
-#'   prototype and/or a size.
-#'
-#' * `vec_assert()` throws an error when the input doesn't conform.
+#' * `vec_is()` is a predicate that checks if its input is a vector that
+#'   conforms to a prototype and/or a size.
+#'
+#' * `vec_assert()` throws an error when the input is not a vector or
+#'   doesn't conform.
+#'
+#' @section Scalars and vectors:
+# From vec_base_typeof():
+#' An object is a vector if one of the following conditions hold:
+#'
+#' - [typeof()] is atomic: `"logical"`, `"integer"`, `"double"`,
+#'   `"complex"`, `"character"`, `"raw"`
+#' - `typeof()` is `"list"`, and one of:
+#'     - the object is a bare `"list"` without a `"class"` attribute,
+#'       see [is_bare_list()]
+#'     - the object is a [data.frame]
+#'     - the `"class"` attribute contains `"list"`, i.e., `inherits(x, "list")`
+#'       is `TRUE`
+#'     - the class implements a [vec_proxy()] method
```

Comment:
This should be moved one level. You can have S3 environments implementing proxy for instance. Non-vector S4 objects will follow the same approach.